### ISSUE TITLE
Fetch promise appears to resolve prematurely when PUT-ing a stream

### ISSUE THREAD

**Issue by universalhandle** [node-fetch/node-fetch#1632]
I'm not sure if I've found a bug in `node-fetch`, or if the problem is in my own code. Apologies if it's the latter, and thanks for your help either way.

Here's my scenario: when triggered, my code should retrieve a WAV from my remote storage location, split it into smaller parts, and upload those to partner endpoints. When all the "part" uploads are complete (but not before), it needs to do other things that are not relevant here. For performance reasons, rather than download the file and chunk it up on the file system, I'd like to make byte-requests for the WAV and hand off the stream from a downloading instance of fetch to an uploading instance of fetch.

Here's some code which demonstrates that the fetch promises are resolving earlier than I expect them to:
```ts
import fetch, { RequestInit } from 'node-fetch';

interface UploadOperation {
  offset: number;
  length: number;
  method: string;
  url: string;
}

function uploadFromUrl(url: string, uploadOperations: UploadOperation[]) {
  const uploadPartPromises = uploadOperations.map(async (op) => {
    const start = op.offset;
    const end = start + op.length;
    const response = await fetch(url, {
      headers: {
        Range: `bytes=${start}-${end}`,
      },
      method: 'GET',
    });
    const stream = response.body;

    stream.on('data', (chunk: string | Buffer) => {
      console.log(
        `Part ${start} to ${end}: ${chunk.length} bytes streamed to uploader`
      );
    });

    const requestInit: RequestInit = {
      body: stream,
      headers: {
        'Content-Length': op.length.toString(),
        'Content-Type': 'audio/wav',
      },
      method: op.method,
    };

    console.log(`Uploading part ${start} to ${end} to ${op.url}`);
    return fetch(op.url, requestInit).finally(() => {
      console.log(`Finished request to ${op.url} for part ${start} to ${end}`);
    });
  });

  return Promise.all(uploadPartPromises).then(() => {
    console.log('All part uploads complete');
    // aforementioned "other stuff" would happen here
  });
}

void uploadFromUrl('https://my-storage.location/path/to/file', [
  {
    offset: 0,
    length: 5242880,
    method: 'PUT',
    url: 'https://partner-storage.location/path/1',
  },
  {
    offset: 5242880,
    length: 5242880,
    method: 'PUT',
    url: 'https://partner-storage.location/path/2',
  },
  {
    offset: 10485760,
    length: 5242880,
    method: 'PUT',
    url: 'https://partner-storage.location/path/3',
  },
]);
```

... and here is abridged output, with commentary:

```
Uploading part 10485760 to 15728640 to https://partner-storage.location/path/3
Part 10485760 to 15728640: 2424 bytes streamed to uploader
Part 10485760 to 15728640: 50540 bytes streamed to uploader
Uploading part 0 to 5242880 to https://partner-storage.location/path/1
Part 0 to 5242880: 2424 bytes streamed to uploader
Part 0 to 5242880: 10640 bytes streamed to uploader
Uploading part 5242880 to 10485760 to https://partner-storage.location/path/2
Part 5242880 to 10485760: 13064 bytes streamed to uploader <!----- by now we can see all three GET bodies are streaming to the PUT commands
[...abridged...]
Finished request to https://partner-storage.location/path/3 for part 10485760 to 15728640
Part 10485760 to 15728640: 2660 bytes streamed to uploader <!---- the promise resolved on the line above, but the GET body is still streaming
[...abridged...]
Finished request to https://partner-storage.location/path/1 for part 0 to 5242880
[...abridged...]
Finished request to https://partner-storage.location/path/2 for part 5242880 to 10485760
All part uploads complete                                  <!----- Promise.all is behaving as expected; it resolves only after its contained promises do (premature though they may be)
Part 10485760 to 15728640: 2660 bytes streamed to uploader <!----- still going
[...abridged...]
Part 0 to 5242880: 30224 bytes streamed to uploader        <!----- still going
[...abridged...]
Part 5242880 to 10485760: 10662 bytes streamed to uploader <!----- still going
[...abridged...]
```

### Environment

| Software        | Version          |
| ----------------- | ------------------ |
| node-fetch     | 2.6.7              |
| node              | 16.4.0            |
| npm 	        | 7.18.1            |
| TypeScript     | 4.7.3              |
| OS                 | Ubuntu 20.04.4 LTS |
